File: src/main/java/org/modelingvalue/nelumbo/When.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import java.io.Serial;

import org.modelingvalue.collections.List;

public final class When extends BinaryPredicate {
    @Serial
    private static final long serialVersionUID = 9105566742523301113L;

    private When(Node when, Node predicate) {
        super(Type.PREDICATE, List.of(), when, predicate);
    }

    private When(Object[] args, When declaration) {
        super(args, declaration);
    }

    public static When of(Node when, Node predicate) {
        return new When(when, predicate);
    }

    @Override
    public When declaration() {
        return (When) super.declaration();
    }

    @Override
    protected When struct(Object[] array, Predicate declaration) {
        return new When(array, (When) declaration);
    }

    @Override
    public When set(int i, Object... a) {
        return (When) super.set(i, a);
    }

    @Override
    protected boolean isTrue(InferResult predResult, int i) {
        return false;
    }

    @Override
    protected boolean isFalse(InferResult predResult, int i) {
        return false;
    }

    @Override
    protected boolean isUnknown(InferResult predResult, int i) {
        return i == 0 && predResult.isFalseCC();
    }

    @Override
    protected boolean isTrue(InferResult[] predResult) {
        return predResult[1].isTrueCC();
    }

    @Override
    protected boolean isFalse(InferResult[] predResult) {
        return predResult[1].isFalseCC();
    }

    @Override
    protected boolean isLeft(InferResult[] predResult) {
        return false;
    }

    @Override
    protected boolean isRight(InferResult[] predResult) {
        return predResult[0].isTrueCC();
    }

    @Override
    protected InferResult add(InferResult[] predResult) {
        return predResult[0].addAnd(predResult[1]);
    }

    @Override
    protected void order(Predicate[] predicate) {
        // Do not change order
    }

    @Override
    public String toString() {
        return predicate2() + " if " + predicate1();
    }
}

File: web-ts/src/logic/When.ts
/**
 * When - conditional predicate (if-then).
 * @JAVA_REF org.modelingvalue.nelumbo.logic.When
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Node } from '../Node';
import { Predicate } from './Predicate';
import { BinaryPredicate } from './BinaryPredicate';
import { InferResult } from '../InferResult';
import { TokenType } from '../syntax/TokenType';

/**
 * When - conditional logic (predicate2 if predicate1).
 * @JAVA_REF org.modelingvalue.nelumbo.logic.When
 */
export class When extends BinaryPredicate {
  constructor(when: Node, predicate: Node) {
    // Java: super(Type.BOOLEAN, List.of(), when, predicate)
    super(Type.BOOLEAN as any, List<AstElement>(), when, predicate);
  }

  protected static fromDataWhen(data: unknown[], declaration?: Node): When {
    const when = Object.create(When.prototype) as When;
    (when as unknown as { _data: unknown[] })._data = data;
    (when as unknown as { _declaration: Node })._declaration = declaration ?? when;
    (when as any)._binding = null;
    (when as any)._hashCodeCached = false;
    (when as any)._hashCode = 0;
    (when as any)._nrOfUnbound = -1;
    return when;
  }

  static of(when: Node, predicate: Node): When {
    return new When(when, predicate);
  }

  protected override struct(data: unknown[], declaration?: Node): When {
    return When.fromDataWhen(data, declaration ?? this.declaration());
  }

  override declaration(): When {
    return super.declaration() as When;
  }

  override set(i: number, ...a: unknown[]): When {
    return super.set(i, ...a) as When;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isTrue(InferResult, int)
  protected isTrueSingle(_predResult: InferResult, _i: number): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isFalse(InferResult, int)
  protected isFalseSingle(_predResult: InferResult, _i: number): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isUnknown(InferResult, int)
  protected isUnknownSingle(predResult: InferResult, i: number): boolean {
    return i === 0 && predResult.isFalseCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isTrue(InferResult[])
  protected isTrueBoth(predResult: InferResult[]): boolean {
    return predResult[0].isTrueCC() && predResult[1].isTrueCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isFalse(InferResult[])
  protected isFalseBoth(predResult: InferResult[]): boolean {
    return predResult[0].isTrueCC() && predResult[1].isFalseCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isLeft(InferResult[])
  protected isLeft(predResult: InferResult[]): boolean {
    return predResult[1].isTrueCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#isRight(InferResult[])
  protected isRight(predResult: InferResult[]): boolean {
    return predResult[0].isTrueCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#resolvedOnly(InferResult[])
  protected override resolvedOnly(predResult: InferResult[]): InferResult {
    if (!predResult[0].unresolvable() && !predResult[1].unresolvable()) {
      return predResult[0].complete().add(predResult[1]);
    } else if (!predResult[0].unresolvable()) {
      return predResult[0].complete();
    } else if (!predResult[1].unresolvable()) {
      return predResult[1];
    } else {
      return InferResult.UNRESOLVABLE;
    }
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.When#toString(TokenType[])
  override toString(previous?: TokenType[]): string {
    return this.predicate2() + ' if ' + this.predicate1();
  }
}
